### ISSUE TITLE
Rework peer connections

### ISSUE THREAD

**Issue by gavofyork** [paritytech/substrate#412]
Avoid the 32-second "try to connect to some nodes" construct.

Instead:
- Split the peer slots into three parts:
  - Reserved (peers that have been passed by the `--reserved` flag - they get their own special slots and should be continuously, though responsibly, (re)connected to)
  - Outgoing (number of slots == `min_peers`, these are connections to nodes that we discover/remember and initiate)
  - Incoming (number of slots == `max_peers - min_peers`, these are inbound connections from nodes that have discovered us)

Regarding Outgoing peers:
- We start by connecting to up to `min_peers` nodes; they each take a slot in our Outgoing peer table.
- We attempt to connect to a (/the highest scoring, once reputation system is done) node when one of the following happens:
  - A previous connection to a node has failed.
  - A peer who was in the Outgoing table has been/become disconnected.
  - We have discovered new nodes and have slots free in the Outgoing peer table.
  - The backoff timer of a previous node has expired and we can retry connecting.

**Comment by arkpar** [comment]
Furthermore certain types of peers should reserve additional slots. E.g. authorities want to be connected to other authorities. The configuration should be flexible enough to allow this.

**Comment by gavofyork** [comment]
yeah, though that's more about general peer steering and should probably go into another issue.

**Comment by tomaka** [comment]
I suggest we eventually rename `min_peers` and `max_peers` to `max_public_outgoing` and `max_public_incoming` (`public` as in "non-reserved"), so that it feels less magic.

**Comment by gavofyork** [comment]
sure
